### ISSUE TITLE
mca_topo_treematch_dist_graph_create hangs

### ISSUE THREAD

**Issue by ggouaillardet** [open-mpi/ompi#2227]
@bosilca can you please have a look at this ?

i run

```
mpirun --mca btl tcp,self --host n1:3,n2:1 -np 4 ./distgraph_test_4
```

from `n1`, and it hangs.
`n1` is a dual socket with 6 cores per socket
`n2` is single socket with 16 cores per socket

here is what happens in `mca_topo_treematch_dist_graph_create`

at line 168, `if(hwloc_bitmap_isincluded(root_obj->cpuset,set))` is `false` on `n1`
(e.g. processes are already bound to a socket) but `true` on `n2`
(e.g. process is seen as unbound, since it is running on the whole and unique socket)
then `check_oversubscribing` returns `true` (since there are 3 tasks on `n1` but only 2 sockets)

then `n1` fallbacks whereas `n2` invokes `check_oversubscribing` again to see if PUs are oversubscribed, and hence the hang.

incidentally, i noted the `FALLBACK()` macro does not reset the `reorder` local variable.

in this case, one task running on a single socket could/should be considered as bound (vs unbound), but even if we do that, there are still some pathes in which the same hang can happen.

the patch below fixes the hang, but is unlikely what we want to do, since it simply avoids trying to bind on PUs when sockets are oversubscribed (note the comments in the code refers sockets as `OBJ/CORE` which is not quite right in this case)

could you please fix the hang or provide some directions on how to move forward ?

``` patch
diff --git a/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c b/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
index 4d4f4d3..52707a4 100644
--- a/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
+++ b/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
@@ -283,24 +283,9 @@ int mca_topo_treematch_dist_graph_create(mca_topo_base_module_t* topo_module,
                                                      nodes_roots,local_procs,comm_old);
         if(oversubscribing_objs) {
 #ifdef __DEBUG__
-            fprintf(stdout,"Oversubscribing OBJ/CORES resources => Trying to use PUs \n");
-#endif
-            int oversubscribed_pus = check_oversubscribing(rank,num_nodes,
-                                                           num_pus_in_node,num_procs_in_node,
-                                                           nodes_roots,local_procs,comm_old);
-            if (oversubscribed_pus){
-#ifdef __DEBUG__
-                fprintf(stdout,"Oversubscribing PUs resources => Rank Reordering Impossible \n");
-#endif
-                FALLBACK();
-            } else {
-                obj_rank = ompi_process_info.my_local_rank%num_pus_in_node;
-                effective_depth = hwloc_topology_get_depth(opal_hwloc_topology) - 1;
-                num_objs_in_node = num_pus_in_node;
-#ifdef __DEBUG__
-                fprintf(stdout,"Process not bound : binding on PU#%i \n",obj_rank);
+            fprintf(stdout,"Oversubscribing OBJ/CORES resources =>  Rank Reordering Impossible\n");
 #endif
-            }
+            FALLBACK();
         } else {
             obj_rank = ompi_process_info.my_local_rank%num_objs_in_node;
             effective_depth = depth;
```

**Comment by bosilca** [comment]
@ggouaillardet thanks for the detailed analysis. Indeed the code made the assumption that the environment was homogeneous. Please give it a try with the following patch

```patch
diff --git a/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c b/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
index 4d4f4d3..9dc1a5f 100644
--- a/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
+++ b/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
@@ -53,7 +53,16 @@
 #define MY_STRING_SIZE 64
 /*#define __DEBUG__ 1  */
 
-
+/**
+ * This function is a allreduce between all processes to detect for oversubscription.
+ * On each node, the local_procs will be a different array, that contains only the
+ * local processes. Thus, that process will compute the node oversubscription and will
+ * bring this value to the operation, while every other process on the node will
+ * contribute 0.
+ * Doing an AllReduce might be an overkill for this situation, but it should remain
+ * more scalable than a star reduction (between the roots of each node (nodes_roots),
+ * followed by a bcast to all processes.
+ */
 static int check_oversubscribing(int rank,
                                  int num_nodes,
                                  int num_objs_in_node,
@@ -62,52 +71,16 @@ static int check_oversubscribing(int rank,
                                  int *local_procs,
                                  ompi_communicator_t *comm_old)
 {
-    int oversubscribed = 0;
-    int local_oversub = 0;
-    int err;
+    int oversubscribed = 0, local_oversub = 0, err;
 
+    /* Only a single process per node, the local root, compute the oversubscription condition */
     if (rank == local_procs[0])
         if(num_objs_in_node < num_procs_in_node)
             local_oversub =  1;
 
-    if (rank == 0) {
-        MPI_Request *reqs = (MPI_Request *)calloc(num_nodes-1, sizeof(MPI_Request));
-        int *oversub = (int *)calloc(num_nodes, sizeof(int));
-        int i;
-
-        oversub[0] = local_oversub;
-        for(i = 1;  i < num_nodes; i++)
-            if (OMPI_SUCCESS != ( err = MCA_PML_CALL(irecv(&oversub[i], 1, MPI_INT,
-                                                           nodes_roots[i], 111, comm_old, &reqs[i-1])))) {
-                /* NTH: more needs to be done to correctly clean up here */
-                free (reqs);
-                free (oversub);
-                return err;
-            }
-
-        if (OMPI_SUCCESS != ( err = ompi_request_wait_all(num_nodes-1,
-                                                          reqs, MPI_STATUSES_IGNORE))) {
-            /* NTH: more needs to be done to correctly clean up here */
-            free (reqs);
-            free (oversub);
-            return err;
-        }
-
-        for(i = 0;  i < num_nodes; i++)
-            oversubscribed += oversub[i];
-
-        free(oversub);
-        free(reqs);
-    } else {
-        if (rank == local_procs[0])
-            if (OMPI_SUCCESS != (err = MCA_PML_CALL(send(&local_oversub, 1, MPI_INT, 0,
-                                                         111, MCA_PML_BASE_SEND_STANDARD, comm_old))))
-                return err;
-    }
 
-    if (OMPI_SUCCESS != (err = comm_old->c_coll.coll_bcast(&oversubscribed, 1,
-                                                           MPI_INT, 0, comm_old,
-                                                           comm_old->c_coll.coll_bcast_module)))
+    if (OMPI_SUCCESS != (err = comm_old->c_coll.coll_allreduce(&local_oversub, &oversubscribed, 1, MPI_INT,
+                                                               MPI_SUM, comm_old, comm_old->c_coll.coll_allreduce_module)))
         return err;
 
     return oversubscribed;
@@ -161,7 +134,7 @@ int mca_topo_treematch_dist_graph_create(mca_topo_base_module_t* topo_module,
     int num_procs_in_node = 0;
     int rank, size;
     int hwloc_err;
-    int oversubscribing_objs = 0;
+    int oversubscribing_objs = 0, oversubscribed_pus = 0;
     int i, j, idx;
     uint32_t val, *pval;
 
@@ -265,8 +238,12 @@ int mca_topo_treematch_dist_graph_create(mca_topo_base_module_t* topo_module,
     hwloc_get_cpubind(opal_hwloc_topology,set,0);
     num_pus_in_node = hwloc_get_nbobjs_by_type(opal_hwloc_topology, HWLOC_OBJ_PU);
 
-    if(hwloc_bitmap_isincluded(root_obj->cpuset,set)){
-        /* processes are not bound on the machine */
+    /**
+     * In all situations (including heterogeneous environments) all processes must execute
+     * all the calls that involve collective communications, so we have to lay the logic
+     * accordingly.
+     */
+    if(hwloc_bitmap_isincluded(root_obj->cpuset,set)){ /* processes are not bound on the machine */
 #ifdef __DEBUG__
         if (0 == rank)
             fprintf(stdout,">>>>>>>>>>>>> Process Not bound <<<<<<<<<<<<<<<\n");
@@ -281,19 +258,35 @@ int mca_topo_treematch_dist_graph_create(mca_topo_base_module_t* topo_module,
         oversubscribing_objs = check_oversubscribing(rank,num_nodes,
                                                      num_objs_in_node,num_procs_in_node,
                                                      nodes_roots,local_procs,comm_old);
-        if(oversubscribing_objs) {
+    } else {    /* the processes are already bound */
+        object = hwloc_get_obj_covering_cpuset(opal_hwloc_topology,set);
+        obj_rank = object->logical_index;
+        effective_depth = object->depth;
+        num_objs_in_node = hwloc_get_nbobjs_by_depth(opal_hwloc_topology, effective_depth);
+
+        /* Check for oversubscribing */
+        oversubscribing_objs = check_oversubscribing(rank,num_nodes,
+                                                     num_objs_in_node,num_procs_in_node,
+                                                     nodes_roots,local_procs,comm_old);
+    }
+
+    if(oversubscribing_objs) {
+        if(hwloc_bitmap_isincluded(root_obj->cpuset,set)){ /* processes are not bound on the machine */
 #ifdef __DEBUG__
             fprintf(stdout,"Oversubscribing OBJ/CORES resources => Trying to use PUs \n");
 #endif
-            int oversubscribed_pus = check_oversubscribing(rank,num_nodes,
-                                                           num_pus_in_node,num_procs_in_node,
-                                                           nodes_roots,local_procs,comm_old);
-            if (oversubscribed_pus){
-#ifdef __DEBUG__
-                fprintf(stdout,"Oversubscribing PUs resources => Rank Reordering Impossible \n");
-#endif
-                FALLBACK();
-            } else {
+            oversubscribed_pus = check_oversubscribing(rank,num_nodes,
+                                                       num_pus_in_node,num_procs_in_node,
+                                                       nodes_roots,local_procs,comm_old);
+        } else {
+            /* Bound processes will participate with the same data as before */
+            oversubscribed_pus = check_oversubscribing(rank,num_nodes,
+                                                       num_objs_in_node,num_procs_in_node,
+                                                       nodes_roots,local_procs,comm_old);
+        }
+        if (!oversubscribed_pus) {
+            /* Update the data used to compute the correct binding */
+            if(hwloc_bitmap_isincluded(root_obj->cpuset,set)){ /* processes are not bound on the machine */
                 obj_rank = ompi_process_info.my_local_rank%num_pus_in_node;
                 effective_depth = hwloc_topology_get_depth(opal_hwloc_topology) - 1;
                 num_objs_in_node = num_pus_in_node;
@@ -301,40 +294,34 @@ int mca_topo_treematch_dist_graph_create(mca_topo_base_module_t* topo_module,
                 fprintf(stdout,"Process not bound : binding on PU#%i \n",obj_rank);
 #endif
             }
-        } else {
-            obj_rank = ompi_process_info.my_local_rank%num_objs_in_node;
-            effective_depth = depth;
-            object = hwloc_get_obj_by_depth(opal_hwloc_topology,effective_depth,obj_rank);
-            if( NULL == object) FALLBACK();
-
-            hwloc_bitmap_copy(set,object->cpuset);
-            hwloc_bitmap_singlify(set); /* we don't want the process to move */
-            hwloc_err = hwloc_set_cpubind(opal_hwloc_topology,set,0);
-            if( -1 == hwloc_err) FALLBACK();
-#ifdef __DEBUG__
-            fprintf(stdout,"Process not bound : binding on OBJ#%i \n",obj_rank);
-#endif
         }
-    } else {    /* the processes are already bound */
-        object = hwloc_get_obj_covering_cpuset(opal_hwloc_topology,set);
-        obj_rank = object->logical_index;
-        effective_depth = object->depth;
-        num_objs_in_node = hwloc_get_nbobjs_by_depth(opal_hwloc_topology, effective_depth);
+    }
 
-        /* Check for oversubscribing */
-        oversubscribing_objs = check_oversubscribing(rank,num_nodes,
-                                                     num_objs_in_node,num_procs_in_node,
-                                                     nodes_roots,local_procs,comm_old);
-        if(oversubscribing_objs) {
+    if( !oversubscribing_objs && !oversubscribed_pus ) {
+        if( hwloc_bitmap_isincluded(root_obj->cpuset,set) ) { /* processes are not bound on the machine */
+        obj_rank = ompi_process_info.my_local_rank%num_objs_in_node;
+        effective_depth = depth;
+        object = hwloc_get_obj_by_depth(opal_hwloc_topology,effective_depth,obj_rank);
+        if( NULL == object) FALLBACK();
+
+        hwloc_bitmap_copy(set,object->cpuset);
+        hwloc_bitmap_singlify(set); /* we don't want the process to move */
+        hwloc_err = hwloc_set_cpubind(opal_hwloc_topology,set,0);
+        if( -1 == hwloc_err) FALLBACK();
+#ifdef __DEBUG__
+        fprintf(stdout,"Process not bound : binding on OBJ#%i \n",obj_rank);
+#endif
+        } else {
 #ifdef __DEBUG__
-            fprintf(stdout,"Oversubscribing OBJ/CORES resources =>  Rank Reordering Impossible\n");
+            fprintf(stdout,"Process %i bound  on OBJ #%i \n",rank,obj_rank);
+            fprintf(stdout,"=====> Num obj in node : %i | num pus in node : %i\n",num_objs_in_node,num_pus_in_node);
 #endif
-            FALLBACK();
         }
+    } else {
 #ifdef __DEBUG__
-        fprintf(stdout,"Process %i bound  on OBJ #%i \n",rank,obj_rank);
-        fprintf(stdout,"=====> Num obj in node : %i | num pus in node : %i\n",num_objs_in_node,num_pus_in_node);
+        fprintf(stdout,"Oversubscribing PUs resources => Rank Reordering Impossible \n");
 #endif
+        FALLBACK();
     }
 
     reqs = (MPI_Request *)calloc(num_procs_in_node-1,sizeof(MPI_Request));
@@ -489,7 +476,6 @@ int mca_topo_treematch_dist_graph_create(mca_topo_base_module_t* topo_module,
                     for(i = 1; i < num_nodes ; i++)
                         displs[i] = displs[i-1] + objs_per_node[i-1];
 
-                    memset(reqs,0,(num_nodes-1)*sizeof(MPI_Request));
                     memcpy(obj_mapping,obj_to_rank_in_comm,objs_per_node[0]*sizeof(int));
                     for(i = 1; i < num_nodes ; i++)
                         if (OMPI_SUCCESS != ( err = MCA_PML_CALL(irecv(obj_mapping + displs[i], objs_per_node[i], MPI_INT,
```

**Comment by ggouaillardet** [comment]
@bosilca i did not have much time today, will try tomorrow

**Comment by ggouaillardet** [comment]
@bosilca 

the patch itself works for me

it evidenced two bugs in other parts of the code
- one was fixed in open-mpi/ompi@e78fcc4db94216f86791035d48c1fda53f1d9188
- the patch below fixes the second one, can you please review it and integrate it in your fix ?
```patch
diff --git a/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c b/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
index 4d4f4d3..2297e5e 100644
--- a/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
+++ b/ompi/mca/topo/treematch/topo_treematch_dist_graph_create.c
@@ -659,7 +659,7 @@ int mca_topo_treematch_dist_graph_create(mca_topo_base_module_t* topo_module,
                 matching = (int *)calloc(size,sizeof(int));
 
                 tm_opt_topology = optimize_topology(tm_topology);
-                comm_tree = build_tree_from_topology(tm_opt_topology,comm_pattern,size,NULL,NULL);
+                comm_tree = build_tree_from_topology(tm_opt_topology,comm_pattern,num_procs_in_node,NULL,NULL);
                 map_topology_simple(tm_opt_topology,comm_tree,matching,size,k);
 #ifdef __DEBUG__
 
```

**Comment by GuillaumeMercier** [comment]
@ggouaillardet 

You second patch does  not seem correct to me , as the matching is computed for all ranks, not just the local processes. Let me try and fix this.

**Comment by GuillaumeMercier** [comment]
@ggouaillardet 

Can you give me details about the bug you encountered when using size instead of num_procs_in_node? Thanks.

**Comment by ggouaillardet** [comment]
I will post the details tomorrow
Basically, build_tree_from_topology is a per node operations, so if size is used instead of num_procs_in_node, and IIRC, treematch sometimes complains there are more processes than constraints.

**Comment by GuillaumeMercier** [comment]
No, it's not always a per node operation. There are two modes: in the centralized mode, one process computes the new reordering and bcasts the result to all processes. In the partially distributed mode,
then it is actually a per node operation as each local root performs the computation. When/if TreeMatch complains, then it's probably a bug.

**Comment by GuillaumeMercier** [comment]
In the centralized mode, the various topologies of the nodes are aggregated to make a single global topology. In such a case, build_tree_from_topology is not per node.

**Comment by GuillaumeMercier** [comment]
@bosilca 
I agree with your patch and will apply it to the new version of the code.

**Comment by bosilca** [comment]
@ggouaillardet and @GuillaumeMercier I will wait until Guillaume have a fix for the second issue to create a PR.

**Comment by GuillaumeMercier** [comment]
@bosilca Sounds sensible ;-)

**Comment by ggouaillardet** [comment]
@GuillaumeMercier 
the logs are at https://gist.github.com/ggouaillardet/f2b2b144808c55074f340774f5fc561e

there are two nodes
- ```n0``` is a dual-socket with 6 cores per socket
- ```n1``` is a single-socket with 16 cores

i ran all cases with
```mpirun --mca btl tcp,self --host n0:$n,n1:$m -np $((n + m)) ./distgraph_test_4```
where ```n=[1...12]``` and ```m=[1...16]```

the first failure occurs when
```mpirun --mca btl tcp,self --host n0:1,n1:7 -np 8 ./distgraph_test_4```

in ```mca_topo_treematch_dist_graph_create``` at line ```471```,
```objs_per_node = {2, 16}```

which is a bit surprising, i would have expected ```{2,1}```

**Comment by GuillaumeMercier** [comment]
@ggouaillardet 
Thanks, I'll take a look at it. It seems that the single process is bound to a socket, not a core,  while the
7 others are bound to cores, hence the {2,16}.

**Comment by ggouaillardet** [comment]
I did not specify any options, so I guess all tasks should be bound to socket.
I will dig more tomorrow, the root cause could be the topology of n1 is missing some levels.

**Comment by GuillaumeMercier** [comment]
If you don't specify any options, then it should be bound to cores by default. The behaviour you're
experiencing is therefore not the one I expect.

**Comment by ggouaillardet** [comment]
I will double check that too, but I am pretty sure default mapping/binding is per socket on master branch

**Comment by GuillaumeMercier** [comment]
Maybe, but if you don't specify anything, processes are bound to cores if cores are available in the dist_graph_create code, not to sockets.

**Comment by ggouaillardet** [comment]
Are we talking about the same binding ?
MPI tasks are bound to resources (e.g. sched_setaffinity via hwloc) by orted/mpirun before they are execv'ed.
topo/treematch enters the game way later (e.g. after MPI_Init if I understand correctly)
Are you saying treematch does the matching as if MPI tasks were bound to cores ?
Or is some glue missing to correctly inform treematch how MPI tasks were bound by hwloc ?

**Comment by GuillaumeMercier** [comment]
There are two bindings : if a binding is enforced by mpirun, the topo/treematch code won't mess with it.
It just checks is such a binding is enforced. If not, then it tries to enforce its own, and it uses cores
as the first target for binding if cores are availables and is oversubscribing is not present.
When you don't specify a binding policy with mpirun, then the processes are not bound, AFAIK.
But then, this might have changed in the actual version. I don't know.

**Comment by GuillaumeMercier** [comment]
@bosilca  George can you confirm the current binding policy of mpirun when nothing is specified?
Are the processes left unbound or are they bound to some ressource?

**Comment by ggouaillardet** [comment]
Let's assume the policy is to bind on socket by default.
If the system is single socket, then a task is bound on all the PUs.
When a task tries to figure out whether it is bound or not, it generally concludes the task is unbound.
I do not think we (can) make a difference between "bound on the unique socket" and "unbound" just by asking hwloc.
Bottom line, per my understanding and your description, topo/treematch messes up with the default bind-to-socket policy on single socket nodes ...
That would explain {2,16}, and/but that might not be the root cause of the error

**Comment by ggouaillardet** [comment]
@GuillaumeMercier you can
mpirun ... grep Cpus_allowed_list /proc/self/status
As long as you are not running on a single socket/core node, you will be able to figure out the default binding policy with that.

**Comment by ggouaillardet** [comment]
On n1 (single socket, 16 cores) what if ?
1. OMP_NUM_THREADS=8 mpirun -np 2 --bind-to none ...
2. OMP_NUM_THREADS=16 mpirun -np 1 --bind-to none ...
In which case does topo/treematch bind the MPI task on a single core ?

**Comment by GuillaumeMercier** [comment]
@ggouaillardet 
Case 1 : each MPI process will be bound to its own core
Case 2  : the single MPI process will be bound to its dedicated core.

I agree that in the case of a single socket  node, it is impossible to make the difference 
(with hwloc) between unbound and bound to the socket.

**Comment by ggouaillardet** [comment]
Case 1 is suboptimsl anyway  (user should ask to bind to 8 cores) but case 2 is problematic.

 should we conclude topo/treematch should never bind MPI tasks at all ?

**Comment by GuillaumeMercier** [comment]
Well, it's problematic because of the OpenMP threads. 
If the process is not bound, then treematch is not able to compute the matching.
But then, the logic of the code can be changed to take this (i.e. OMP threads) into account
(don't know how yet ...)
Or in case 2, if the user asks to bind to socket, then it should be ok, except that we don't catch
the case of the single socket per node case.  A possible fix would be to check that a mpirun
binding is enforced and then respect it instead of trying to detect if a binding policy has been
asked.

**Comment by GuillaumeMercier** [comment]
@bosilca : is there a way to have the information that a binding has been asked by the user
with mpirun?

**Comment by bosilca** [comment]
@rhc54 how can we figure out from the MPI library if the processes are subject to a user requested binding ?

**Comment by rhc54** [comment]
It's a timely question - we are just starting a PMIx working group to discuss this very question of coordinating RM/OpenMP/MPI binding requests, including dynamic adjustments. Some prelim meetings are being held at SC this coming week, and we'll kick things off more actively after the holiday.

In the interim, the variable ```orte_proc_is_bound``` will be true if the process is bound in some way. The environmental variable ```OMPI_MCA_orte_bound_at_launch``` indicates that mpirun bound the process. The environmental variable ```OMPI_MCA_orte_externally_bound``` gets set if SLURM defined the binding.

We don't have binding detection for other RM's just yet - hopefully, the PMIx working group will help fill that gap.

**Comment by GuillaumeMercier** [comment]
@rhc54 Thanks, I'll use this for now.

**Comment by GuillaumeMercier** [comment]
A question : is the variable orte_proc_is_bound still usable for this purpose?
 Because on my one socket machine,  when i don't bind my processes I've got this variable set to true and --display-map says:   Bound: N/A
When I bind to the single socket (and the machine in fact), it is still set to true but --display-map says: 
Bound: UNBOUND
So there seem to be a distinction made but this variable is not the right way to catch it.

**Comment by bosilca** [comment]
@rhc54 I can confirm that orte_proc_is_bound seems to always be set, even when I expressly request "--bind-to none". If I add '--report-bindings --display-map" I see what I expect: the processes are not bound. 
```txt
 Data for JOB [58330,1] offset 0

 ========================   JOB MAP   ========================

 Data for node: a00	Num slots: 1	Max slots: 0	Num procs: 1
 	Process OMPI jobid: [58330,1] App: 0 Process rank: 0 Bound: N/A

 Data for node: a01	Num slots: 7	Max slots: 0	Num procs: 7
 	Process OMPI jobid: [58330,1] App: 0 Process rank: 1 Bound: N/A
 	Process OMPI jobid: [58330,1] App: 0 Process rank: 2 Bound: N/A
 	Process OMPI jobid: [58330,1] App: 0 Process rank: 3 Bound: N/A
 	Process OMPI jobid: [58330,1] App: 0 Process rank: 4 Bound: N/A
 	Process OMPI jobid: [58330,1] App: 0 Process rank: 5 Bound: N/A
 	Process OMPI jobid: [58330,1] App: 0 Process rank: 6 Bound: N/A
 	Process OMPI jobid: [58330,1] App: 0 Process rank: 7 Bound: N/A

 =============================================================
[a00:33624] MCW rank 0 is not bound (or bound to all available processors)
[a01:32847] MCW rank 1 is not bound (or bound to all available processors)
[a01:32849] MCW rank 3 is not bound (or bound to all available processors)
[a01:32850] MCW rank 4 is not bound (or bound to all available processors)
[a01:32851] MCW rank 5 is not bound (or bound to all available processors)
[a01:32852] MCW rank 6 is not bound (or bound to all available processors)
[a01:32853] MCW rank 7 is not bound (or bound to all available processors)
[a01:32848] MCW rank 2 is not bound (or bound to all available processors)
```

**Comment by rhc54** [comment]
is this on master?

**Comment by bosilca** [comment]
On my fork. I sync it with OMPI master yesterday so it should be pretty much up-to-date.

**Comment by rhc54** [comment]
Got it - PR in progress

**Comment by rhc54** [comment]
See https://github.com/open-mpi/ompi/pull/3699

**Comment by bosilca** [comment]
Thanks @rhc54. I'll check it right away.

**Comment by GuillaumeMercier** [comment]
That leaves the question open: how does ORTE make the difference between unbound and N/A statuses?

**Comment by rhc54** [comment]
What do you mean by "N/A"? You are either bound or not bound, yes?

**Comment by GuillaumeMercier** [comment]
Ok, let me specify what I meant: when running a 2-process app on my single-socket machine, when I use the --bind-to socket option , the process manager outputs: 
Process OMPI jobid: [38185,1] App: 0 Process rank: 0 Bound: N/A
Process OMPI jobid: [38185,1] App: 0 Process rank: 1 Bound: N/A
But when I use --bind-to socket it outputs:
Process OMPI jobid: [38145,1] App: 0 Process rank: 0 Bound: UNBOUND
Process OMPI jobid: [38145,1] App: 0 Process rank: 1 Bound: UNBOUND

So I guess at some point a disctinction is made and that's what I'd like to know because to me, there is a difference in such case.

**Comment by rhc54** [comment]
I believe the difference is that, in the first case, you specifically asked us to ignore binding, and so the "bound" flag indicates it isn't applicable. In the second case, we are supposed to bind to socket, but there is only one socket - thus, it is reporting "unbound" because the proc is actually bound to every available processor on the node, which is the same thing as being "unbound".

The result in both cases is identical - we haven't "bound" the proc to anything.

**Comment by GuillaumeMercier** [comment]
no, there is a difference to me:  in the case where the user did not specify a binding, I can then enforce a binding of my own. In the other case, the user specified something: maybe he will create an OpenMP parallel section that will need the cores. But in this case, I should not enforce ant binding whatsoever. And I'd like to handle both cases correctly, hence my question.

**Comment by ggouaillardet** [comment]
what do you mean by "the user does not specify a binding" ?
do you mean ```--bind-to none``` ?

if you are running on a single socket, ```--bind-to none``` and ```--bind-to socket``` will have the same behavior (not binding equals binding to all cores) but it might be reported differently (e.g. ```N/A``` vs ```UNBOUND```)

i am not sure of what you are trying to achieve and how, but you might want to query a MCA parameter and check how it was set (command line, environment, config file or unset)

**Comment by GuillaumeMercier** [comment]
Yes, I mean --bind-to none. I agree that on a single socket, --bind-to none and --bind-to socket will have the same behaviour but the fact is that the user does not specify the same thing and that's I want to handle. Actually, I'd like to handle the case that you reported to me months ago.  I tried to check  the orte_proc_is_bound variable but it seems to be set to TRUE regardless of the situation. And I'm curious to know how the process manager makes the difference between N/A and UNBOUND.

**Comment by bosilca** [comment]
This issue has been addressed in 7a866f754c3
